bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ManagedThingProvider.java

```diff
@@ -52,4 +80,158 @@ protected void unsetStorageService(StorageService storageService) {
+            if (!aAcceptedItemType.equals(bAcceptedItemType)) {
+                return false;
+            }
+        } else if (bAcceptedItemType != null) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private void createThingFromStorageForThingHandlerFactory(Thing storedThing, ThingHandlerFactory factory) {
+        if (!loadedXmlThingTypes.contains(getBundleName(factory))) {
+            return;
+        }
+
+        if (factory.supportsThingType(storedThing.getThingTypeUID())) {
+            ThingBuilder thingBuilder = ThingBuilder.create(storedThing.getThingTypeUID(), storedThing.getUID())
+                    .withBridge(storedThing.getBridgeUID()).withChannels(storedThing.getChannels())
+                    .withConfiguration(storedThing.getConfiguration()).withLabel(storedThing.getLabel())
+                    .withLocation(storedThing.getLocation()).withProperties(storedThing.getProperties());
+
+            Configuration storedConfiguration = storedThing.getConfiguration();
+            Thing factoryThing = factory.createThing(storedThing.getThingTypeUID(), storedConfiguration,
+                    storedThing.getUID(), storedThing.getBridgeUID());
+
+            if (factoryThing != null) {
+                Map<ChannelUID, Channel> factoryChannels = factoryThing.getChannels().stream()
+                        .collect(Collectors.toMap(Channel::getUID, channel -> channel));
+                Map<ChannelUID, Channel> storedChannels = storedThing.getChannels().stream()
+                        .collect(Collectors.toMap(Channel::getUID, channel -> channel));
+
+                // check if we have new or updated channels
+                factoryChannels.forEach((factoryChannelUID, factoryChannel) -> {
+                    if (!storedChannels.containsKey(factoryChannelUID)) {
+                        thingBuilder.withChannel(factoryChannel);
+                        logger.trace("added channel {} to thing {}", factoryChannel, storedThing.getUID());
+                    } else {
+                        if (!compareChannels(factoryChannel, storedThing.getChannel(factoryChannelUID.getId()))) {
+                            thingBuilder.withoutChannel(factoryChannelUID).withChannel(factoryChannel);
+                            logger.trace("updated channel {} in thing {}", factoryChannel, storedThing.getUID());
+                        }
+                    }
+                });
+
+                // check if we have removed channels
+                storedChannels.forEach((storedChannelUID, storedChannel) -> {
+                    if (!factoryChannels.containsKey(storedChannelUID)) {
+                        logger.trace("channel {} in thing {} no longer present in XML", storedChannel,
+                                storedThing.getUID());
+                    }
+                });
+            }
+            Thing newThing = thingBuilder.build();
+
+            if (thingsList.contains(storedThing)) {
```

Comment:
when will this check hold true? New things are only added to the `thingsList` in the `else` case in line 226. So the storedThing is not supposed to be in there. Besides that it should also be exchanged by the `newThing` entity.